odl/test/space/pspace_test.py

```diff
@@ -900,5 +900,110 @@ def test_reductions():
+    x = noise_element(space)
+
+    # Test real
+    expected_result = space.real_space.element([x[0].real, x[1].real])
+    assert x.real == expected_result
+
+    # Test imag
+    expected_result = space.real_space.element([x[0].imag, x[1].imag])
+    assert x.imag == expected_result
+
+    # Test conj. Note that ProductSpace does not implement asarray if
+    # is_power_space is false. Hence the construction below
+    expected_result = space.element([x[0].conj(), x[1].conj()])
+    x_conj = x.conj()
+    assert x_conj[0] == expected_result[0]
+    assert x_conj[1] == expected_result[1]
+
+
+def test_real_setter():
+    """Verify that the setter for the real part of an element works."""
+    space = odl.ProductSpace(odl.uniform_discr(0, 1, 3, dtype=complex),
+                             odl.cn(2))
+    x = noise_element(space)
+    x_imag = x.imag
+
+    # Test for element in ProductSpace where imaginary part is zero
+    tmp = noise_element(space)
+    newreal = space.element(tmp.real)
+    x.real = newreal
+    assert x.real == newreal.real
+
+    # Test for element in ProductSpace with none-zero imaginary part
+    newreal = noise_element(space)
+    x.real = newreal
+    assert x.real == newreal.real
+
+    # Test for element in ProductSpace.real_space
+    tmp = noise_element(space)
+    newreal = tmp.real
+    x.real = newreal
+    assert x in space
+    assert x.real == newreal
+
+    # Test for array of numpy.ndarray
+    tmp = noise_element(space)
+    newreal = [tmp[0].real.asarray(), tmp[1].real.asarray()]
+    x.real = newreal
+    expected_result = space.real_space.element(newreal)
+    assert x.real == expected_result
+
+    # Test array of numpy.ndarray with none-zero imaginary part
+    tmp = noise_element(space)
+    newreal = [tmp[0].asarray(), tmp[1].asarray()]
+    x.real = newreal
+    expected_result = space.real_space.element(newreal)
+    assert x.real == expected_result
+
+    # Test for array
+    newreal = [[0, 1, 2], [3, 4]]
+    x.real = newreal
+    expected_result = space.real_space.element(newreal)
+    assert x.real == expected_result
+
+    # Test array with none-zero imaginary part
+    newreal = [[0, 1 + 1j, 2], [3, 4]]
+    x.real = newreal
+    expected_result = space.element(newreal).real
+    assert x.real == expected_result
+
+    # Test for scalar
+    newreal = np.random.randn()
+    x.real = newreal
+    expected_result = (newreal * space.one()).real
+    assert x.real == expected_result
+
+    # Test scalar with none-zero imaginary part
+    newreal = np.random.randn() + 1j * np.random.randn()
+    x.real = newreal
+    expected_result = (newreal * space.one()).real
+    assert x.real == expected_result
+
+    # Assert that imaginary part has not changed in any of the tests
+    assert x.imag == x_imag
+
+    # Test that wrong first dimension do not pass
+    newreal = np.random.randn(5, 1)  # space has 5 entries in total
+    with pytest.raises(ValueError):
+        x.real = newreal
```

Comment:
Can you parametrize this test? It's a lot of repetition, I think it wouldn't be too hard to make a fixture for this.

Also, there's no test for the power space case with broadcasting yet, that would be an important one.